dmarc/lookup.go

```diff
@@ -51,13 +51,54 @@ func LookupWithOptions(domain string, options *LookupOptions) (*Record, error) {
 		}
 		return nil, errors.New("dmarc: failed to lookup TXT record: " + err.Error())
 	}
+
+	// net.LookupTXT will concatenate strings contained in a single TXT record.
+	// In other words, net.LookupTXT returns one entry per TXT record, even if
+	// a record contains multiple strings.
 	if len(txts) == 0 {
 		return nil, ErrNoPolicy
 	}
 
-	// Long keys are split in multiple parts
-	txt := strings.Join(txts, "")
-	return Parse(txt)
+	// RFC 6376:
+	// Records that do not start with a "v=" tag that identifies the
+	// current version of DMARC are discarded.
+	dmarcRecords := make([]string, 0, len(txts))
+	for _, record := range txts {
+		if IsDmarcRecord(record) {
+			dmarcRecords = append(dmarcRecords, record)
+		}
+	}
+
+	if len(dmarcRecords) != 1 {
+		return nil, ErrNoPolicy
+	}
+
+	return Parse(dmarcRecords[0])
+}
+
+func IsDmarcRecord(txt string) bool {
```

Comment:
According to [section 6.6.3](https://datatracker.ietf.org/doc/html/rfc7489#section-6.6.3), checking for the `v=` prefix should be enough.

Reply:
But it states: **that identifies the current version of DMARC**
There is also:
```
   v: Version (plain-text; REQUIRED).  Identifies the record retrieved
      as a DMARC record.  It MUST have the value of "DMARC1".  The value
      of this tag MUST match precisely; if it does not or it is absent,
      the entire retrieved record MUST be ignored.  It MUST be the first
      tag in the list.
```

I believe that _the current version_ is part of the statement as if there was a conjunction.

Reply:
Also, matching only the `v=` would match SPF and DKIM records. That wouldn't make much sense.

Reply:
> that identifies the current version of DMARC

Hm, you're right. I'm not a fan of duplicating the parsing logic though. I'd suggest either matching on the `v=DMARC1` prefix, or parsing all records for key-values.

> Also, matching only the v= would match SPF and DKIM records. That wouldn't make much sense.

DMARC records are stored in a separate, special "_dmarc" subdomain, so that shouldn't be the case.

Reply:
> Hm, you're right. I'm not a fan of duplicating the parsing logic though. I'd suggest either matching on the `v=DMARC1` prefix, or parsing all records for key-values.

That would also match things like `v=DMARC1.0` or `v=DMARC10` that shouldn't be matched. We also can't simple match `v=DMARC1;` as leading and trailing spaces in the tag value should be ignored. I believe that the current logic is quite robust and makes sense.

Reply:
@emersion I'm not quite sure how I should change the pull request. Could you please propose something more specific?

Reply:
The current logic duplicates parsing, and I'd like to avoid this. `parseParams()` can be called instead.

Reply:
I removed the duplication now.